pkg/sources/docker/docker.go

```diff
@@ -71,164 +72,202 @@ func (s *Source) Init(_ context.Context, name string, jobId, sourceId int64, ver
 	return nil
 }
 
+type imageInfo struct {
+	image v1.Image
+	base  string
+	tag   string
+}
+
+type layerInfo struct {
+	digest v1.Hash
+	base   string
+	tag    string
+}
+
 // Chunks emits data over a channel that is decoded and scanned for secrets.
 func (s *Source) Chunks(ctx context.Context, chunksChan chan *sources.Chunk) error {
-	remoteOpts, err := s.remoteOpts()
-	if err != nil {
-		return err
-	}
+	ctx = context.WithValues(ctx, "source_type", s.Type(), "source_name", s.name)
 
-	logger := ctx.Logger().WithValues("source_type", s.Type(), "source_name", s.name)
+	workers := new(errgroup.Group)
+	workers.SetLimit(s.concurrency)
 
 	for _, image := range s.conn.GetImages() {
-		var img v1.Image
-		var err error
-		var base, tag string
-		var hasDigest bool
-		var imageName name.Reference
-
-		if strings.HasPrefix(image, "file://") {
-			image = strings.TrimPrefix(image, "file://")
-			base = image
-			img, err = tarball.ImageFromPath(image, nil)
+		image := image
+		workers.Go(func() error {
+			if common.IsDone(ctx) {
+				return nil
+			}
+
+			imgInfo, err := s.processImage(ctx, image)
 			if err != nil {
 				return err
 			}
-		} else {
-			base, tag, hasDigest = baseAndTagFromImage(image)
 
-			if hasDigest {
-				imageName, err = name.NewDigest(image)
-				if err != nil {
-					return err
-				}
-			} else {
-				imageName, err = name.NewTag(image)
-				if err != nil {
-					return err
-				}
-			}
+			ctx = context.WithValues(ctx, "image", imgInfo.base, "tag", imgInfo.tag)
+			ctx.Logger().V(2).Info("scanning image")
 
-			img, err = remote.Image(imageName, remoteOpts...)
+			layers, err := imgInfo.image.Layers()
 			if err != nil {
 				return err
 			}
-		}
 
-		logger = logger.WithValues("image", base, "tag", tag)
-		logger.V(2).Info("scanning image")
+			for _, layer := range layers {
+				if err := s.processLayer(ctx, layer, imgInfo, chunksChan); err != nil {
+					return err
+				}
+				dockerLayersScanned.WithLabelValues(s.name).Inc()
```

Comment:
you also increment this inside `processLayer` - was that on purpose?

Reply:
whoops. Removed the increment in processLayer.